pkg/search/controller.go

```diff
@@ -161,46 +163,210 @@ func (c *Controller) handleErr(err error, key interface{}) {
+	} else {
+		klog.V(4).Infof("No registries added to cluster, cluster: %s", cluster)
+	}
+	if len(removedRegistries) > 0 {
+		klog.V(4).Infof("Old registries removed from cluster, cluster: %s, registries: %s", cluster, strings.Join(removedRegistries, ", "))
+	} else {
+		klog.V(4).Infof("No registries removed from cluster, cluster: %s", cluster)
+	}
+	var addedResources, removedResources []schema.GroupVersionResource
+	cr, addedResources, removedResources = c.getClusterRegistriesModification(allRegistries, matchedRegistries, cr, addedRegistries, removedRegistries)
+	currentWatchingResources := util.Keys(cr.resources)
+	resourcesString := util.StringerJoin(currentWatchingResources, ", ")
+	if len(addedResources) > 0 {
+		klog.V(4).Infof("New watching resources added to cluster, cluster: %s, resources: %s", cluster, util.StringerJoin(addedResources, ", "))
+	} else {
+		klog.V(4).Infof("No watching resources added to cluster, cluster: %s, resources: %s", cluster, resourcesString)
+	}
+	if len(removedResources) > 0 {
+		klog.V(4).Infof("Resources watching removed from cluster, cluster: %s, resources: %s", cluster, util.StringerJoin(removedResources, ", "))
+	} else {
+		klog.V(4).Infof("No resources watching removed from cluster, cluster: %s, resources: %s", cluster, resourcesString)
+	}
+	resourcesChanged = len(addedResources) > 0 || len(removedResources) > 0
+	newRegistry = !hasExistingRegistries
+	return
+}
+
+func (c *Controller) getRegistryBackendHandler(cluster string, matchedRegistries map[string]*searchv1alpha1.ResourceRegistry) (cache.ResourceEventHandler, error) {
+	backend := backendstore.GetBackend(cluster)
+	if backend == nil {
+		backendConfig := c.findOpenSearchBackend(matchedRegistries)
+		backendstore.AddBackend(cluster, backendConfig)
+		backend = backendstore.GetBackend(cluster)
+	}
+	if backend == nil {
+		return nil, fmt.Errorf("failed to get backend store for cluster %s", cluster)
+	}
+	handler := backend.ResourceEventHandlerFuncs()
+	if handler == nil {
+		return nil, fmt.Errorf("failed to get resource event handler for cluster %s", cluster)
+	}
+	return handler, nil
+}
+
+// doCacheCluster processes the resourceRegistry object
+// TODO: update status
+func (c *Controller) doCacheCluster(cluster string) error {
+	// STEP0: stop informer manager for the cluster which does not exist anymore or is not ready.
+	cls, able, err := c.clusterAbleToCache(cluster)
+	if err != nil || !able {
+		return err
+	}
+	matchedRegistries, cr, resourcesChanged, newRegistry, err := c.reconcileClusterWithRegistries(cls)
+	if err != nil {
+		return err
+	}
+
+	// STEP1:  stop informer manager for the cluster which is not referenced by any `SearchRegistry` object.
+	if cr.unregistry() {
+		klog.Infof("Try to stop cluster informer %s", cluster)
+		c.InformerManager.Stop(cluster)
 		return nil
 	}
 
+	if resourcesChanged {
+		klog.Infof("Create new informer for cluster %s", cluster)
+		c.InformerManager.Stop(cluster)
+	}
```

Comment:
Do we need to stop the informer before making changes to the resources, or can we directly modify the current informer?

Reply:
No, like the comment above, I firstly came the same idea as you, so I did `diff` to find out what are added and what are removed. But it just not work.

You can find the first commit of this PR, I did the exact same thing, modifing the current informer.

Reply:
https://github.com/karmada-io/karmada/blob/a9df8656a933d28e9f42e5c1a816ea9c9c1bde32/pkg/util/fedinformer/genericmanager/single-cluster-manager.go#L19-L22

Here, the first commit of this PR, I tried to add a `UnwatchResource(resource schema.GroupVersionResource)`, without success.

Reply:
ok, thx~